scripts/kc_data_transfer/src/CsvTransfer.py

```diff
@@ -0,0 +1,87 @@
+import csv
+import logging
+from KeyCloakTransfer import FamWrapper
+import constants
+
+LOGGER = logging.getLogger(__name__)
+
+
+class AppToFam:
+    def __init__(self):
+        self.path = constants.APP_USER_FILE_PATH
+        self.fam = FamWrapper()
+
+    def get_forest_client(self, organization: str):
+        # get the forest client number from the "For Organization" column
+        # if has one, it should be in the format of
+        # "[forest_client_number] - [organization_name]"
+        forest_client_num = organization.split('-')[0].replace(" ", "")
+        if forest_client_num.isdigit():
+            return forest_client_num
+        return None
+
+    def get_username_and_type(self, user: str):
+        # get username and type from the "User" column in the csv file
+        # it should be in the format of "USERSTYPE\USERNAME"
+        user_info = user.split('\\')
+        if len(user_info) == 2:
+            if "IDIR" in user_info[0]:
+                return "I", user_info[1]
+            if "BCEID" in user_info[0]:
+                return "B", user_info[1]
+
+        return None, None
+
+    def transfer_user_to_fam(self):
+        app_id = self.fam.get_fom_app_id()
+        users = self.read_file_data(self.path)
+
+        for index, user in enumerate(users):
+            user_type_code, username = self.get_username_and_type(user['User'])
+            if user_type_code and username:
+                role_name = user["Profile"]
+                role_id = self.fam.get_role_id(role_name, app_id)
+
+                if role_id:
+                    print(index, username, user_type_code, role_id,
+                          self.get_forest_client(user["For Organization"]))
+                    self.fam.create_user_role_assignment(
+                        user_type_code=user_type_code,
+                        user_name=username,
+                        role_id=role_id,
+                        role_type_code="C",
+                        forest_client_number=self.get_forest_client(
+                            user["For Organization"]
+                        ),
+                    )
+                else:
+                    LOGGER.error("failed to find role, " +
+                                 f"please check the data format: {user}")
+            else:
+                LOGGER.error("failed to get user name and user type, " +
+                             f"please check the data format: {user}")
+
+    def read_file_data(self, path: str):
+        data = []
+        column_title = []
+
+        with open(path, newline='', encoding='utf-8-sig') as csvfile:
+            file_content = csv.reader(csvfile, delimiter=',')
```

Comment:
Would it be easier to use `csv.DictReader(csvfile)` instead? It directly translates csv into dictionary for you without you to construct your dataset. It could be problematic to parse data your own (for example, if data row columns, does not match title row columns, or index).
The [Python csv documentation](https://docs.python.org/3.8/library/csv.html#csv.reader) has this example that is easy to understand:
```
>>> with open('names.csv', newline='') as csvfile:
...     reader = csv.DictReader(csvfile)
...     for row in reader:
...         print(row['first_name'], row['last_name'])
...
Eric Idle
John Cleese

>>> print(row)
{'first_name': 'John', 'last_name': 'Cleese'}
```

Reply:
I tried this method, this one requires we do everything while the csv file is open, so we need to write everything under "with open('names.csv', newline='') as csvfile:", or copy the the whole DictReader data, I feel it's okay to leave it as what we have now

Reply:
Hm... not really sure what you meant... since both method open a file for readying every row...; it is just instead of doing custom row parsing by the column index into a data set, we take advantage on cleaner approach using the second method with just using the "titles" and does not rely on specific column "index". We don't really need to write everything under the "open()". Assigning user role can still be out of the open() loop.

Reply:
yeah, it's just that currently we call the "read_file_data" method to get all user data, where the "read_file_data" method read the file and return the data.  But if I use "csv.DictReader", then it must keep the file open while go through the data, so we can't have "read_file_data" as a separate function, everything need to be done while the csv file open, so like:
"""
def transfer_user_to_fam(self):
        app_id = self.fam.get_app_id()

        with open('names.csv', newline='') as csvfile:
            users = csv.DictReader(csvfile)
            for user in users:
                .....line39 - line64
"""
If I just changed the "read_file_data" function to use "csv.DictReader", it complains that "ValueError: I/O operation on closed file."

<img width="645" alt="Screen Shot 2023-04-25 at 12 37 41 PM" src="https://user-images.githubusercontent.com/23131735/234384755-b6559e87-3736-4792-811d-c1d5ac302f92.png">